src/consensus/vote.cpp

```diff
@@ -570,6 +614,82 @@ std::string VoteManager::getJsonStr(std::vector<Vote> const& votes) {
   return ptroot.toStyledString();
 }
 
+std::vector<std::shared_ptr<Vote>> VoteManager::getProposalVotes(uint64_t pbft_round) {
+  std::vector<std::shared_ptr<Vote>> proposal_votes;
+
+  sharedLock_ lock(verified_votes_access_);
+  for (auto const& voted_value : verified_votes_[pbft_round][1]) {
+    for (auto const& v : voted_value.second) {
+      proposal_votes.emplace_back(std::make_shared<Vote>(v.second));
+    }
+  }
+
+  return proposal_votes;
+}
+
+VotesBundle VoteManager::getVotesBundleByRoundAndStep(uint64_t round, size_t step, size_t two_t_plus_one) {
+  std::vector<Vote> votes;
+
+  sharedLock_ lock(verified_votes_access_);
+  auto found_round_it = verified_votes_.find(round);
+  if (found_round_it != verified_votes_.end()) {
+    auto found_step_it = found_round_it->second.find(step);
+    if (found_step_it != found_round_it->second.end()) {
+      for (auto const& voted_value : found_step_it->second) {
+        if (voted_value.second.size() >= two_t_plus_one) {
+          auto voted_block_hash = voted_value.first;
+          votes.reserve(two_t_plus_one);
+
+          auto it = voted_value.second.begin();
+          size_t count = 0;
+          // Only copy 2t+1 votes
+          while (count != two_t_plus_one) {
+            votes.emplace_back(it->second);
+            it++;
+            count++;
+          }
+          LOG(log_nf_) << "Found enough " << votes.size() << " votes at voted value " << voted_block_hash
+                       << " for round " << round << " step " << step;
+          return VotesBundle(true, voted_block_hash, votes);
```

Comment:
again you are copying the whole votes vector here, which will copy all the votes inside. Can you please go over you code (just to make sure) that you are using move anywhere you can to avoid useless copies like this ?